### PR TITLE
[QNN EP] Update QNN SDK to version 2.17.0

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
- Update QNN CI Pipelines to use QNN SDK version 2.17.0
- **Print warning if unit test requires adjusted tolerance to pass**
- **Temporarily disable unloading QnnCpu.dll for windows x64 due to crash when calling FreeLibrary**
- Enable fixed HTP tests
  - QnnHTPBackendTests.LayerNorm1D_LastAxis_DynamicScale
  - QnnHTPBackendTests.GlobalMaxPool_LargeInput2_u8
  - QnnHTPBackendTests.ReduceSumS8Opset13_Rank5
  - QnnHTPBackendTests.ReduceSumU8Opset13_Rank5_LastAxis
  - QnnHTPBackendTests.WhereLargeDataBroadcastU8
  - QnnHTPBackendTests.WhereLargeDataBroadcastTransformedU8
- Enabled fixed CPU tests
  - QnnCPUBackendTests.Resize_DownSample_Linear_AlignCorners_scales
- Increased tolerance for HTP tests that are less accurate on QNN SDK 2.17.0
  - QnnHTPBackendTests.AveragePool_CountIncludePad_HTP_u8
  - QnnHTPBackendTests.AveragePool_AutopadSameUpper_HTP_u8
  - QnnHTPBackendTests.AveragePool_AutopadSameLower_HTP_u8
  - QnnHTPBackendTests.ConvU8U8S32_bias_dynamic_input
  - QnnHTPBackendTests.ConvU8U8S32_bias_initializer
  - QnnHTPBackendTests.ConvU8U8S32_large_input1_padding_bias_initializer
  - QnnHTPBackendTests.LRNSize3
  - QnnHTPBackendTests.LRNSize5
  - QnnHTPBackendTests.MaxPool_Large_Input_HTP_u8
  - QnnHTPBackendTests.MaxPool_LargeInput_1Pads
  - QnnHTPBackendTests.Resize_DownSample_Linear_HalfPixel
  - QnnHTPBackendTests.ResizeU8_2xLinearPytorchHalfPixel
  - QnnHTPBackendTests.ResizeU8_2xLinearHalfPixel
  - QnnHTPBackendTests.ResizeU8_2xLinearAlignCorners
  - QnnHTPBackendTests.ResizeU8_2xLinearAsymmetric
- Disabled ONNX model tests
  - averagepool_2d_ceil: Accuracy issues **only on Windows x64 QnnCpu.dll**
- Disabled QDQ model tests (onnx_test_runner)
  - facedetection_op8_qdq: Accuracy issues
- Disabled CPU EP tests (these use QnnCpu.dll)
  - ActivationOpTest.Relu: QNN SDK 2.17 Relu treats inf as FLT_MAX
  - GemmOpTypedTests/0.TestGemmBroadcast: Inaccuracy when weight is initializer and bias is not
  - MathOpTest.MatMulFloatType "test padding and broadcast B > A": Inaccuracy (**only linux**)
- Fix Gemm translation bugs in QNN EP:
  - Do not skip processing of inputs that need to be transposed.

### Motivation and Context
- Allow testing with newest QNN SDK version
- Take advantage of improvements to enable new models.

### CODE REVIEW

File: onnxruntime/test/providers/qnn/qnn_test_utils.h [hunk 1]
```diff
@@ -366,37 +381,69 @@ inline void TestQDQModelAccuracy(const GetTestModelFn& f32_model_fn, const GetTe
...
+              << "abs(qdq@QNN_EP - qdq@CPU_EP) / output_range = " << qdq_vals_err_norm * 100.0f << "%";
+
+          max_f32_err = std::max(max_f32_err, qnn_err_norm);
+          max_qdq_err = std::max(max_qdq_err, qdq_vals_err_norm);
+          if (passed_test && !is_as_accurate_as_cpu_ep && (qdq_vals_err_norm > QDQTolerance::DEFAULT_QDQ_TOLERANCE)) {
+            print_accuracy_warning = true;
+          }
+        }
+
+        if (print_accuracy_warning) {
```
Comment (adrianlizarraga) [inline onnxruntime/test/providers/qnn/qnn_test_utils.h:438]: Note: print warning if we run a test needs a larger tolerance to pass.

File: onnxruntime/core/providers/qnn/builder/opbuilder/gemm_op_builder.cc [hunk 2]
```diff
@@ -92,7 +92,10 @@ Status GemmOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
     utils::InitializeQuantizeParam(quantize_param, is_quantized_tensor);
 
     const auto& input_name = inputs[input_i].node_arg.Name();
-    if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_name)) {
+
+    // Only skip if the input tensor has already been added (by producer op) *and* we don't need
+    // to transpose it.
+    if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_name) && input_trans_flag[input_i] == 0) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/gemm_op_builder.cc:98]: Note: This bug primarily prevented Gemm with dynamic activations (that are not graph inputs) from passing qnn validation.

File: onnxruntime/test/providers/cpu/activation/activation_op_test.h [hunk 3]
```diff
@@ -46,11 +46,12 @@ inline void TestActivationOp(const char* szOp, const std::vector<std::vector<T>>
...
 #endif
 
-// Disabled because of NNAPI treat float::inf as float::max
-#if defined(USE_NNAPI)
+// Disabled because NNAPI and QNN EP (SDK 2.17) treat float::inf as float::max
+#if defined(USE_NNAPI) || defined(USE_QNN)
     int relu = strcmp(szOp, "Relu");
     if (relu == 0) {
       excluded_providers.insert(kNnapiExecutionProvider);
+      excluded_providers.insert(kQnnExecutionProvider);
```
Comment (adrianlizarraga) [inline onnxruntime/test/providers/cpu/activation/activation_op_test.h:54]: Note: added the test `QnnCPUBackendTests.DISABLED_UnaryOp_Relu` with a note to enable here when fixed.

File: onnxruntime/test/providers/cpu/math/gemm_test.cc [hunk 4]
```diff
@@ -357,10 +357,19 @@ TYPED_TEST(GemmOpTypedTests, TestGemmBroadcast) {
...
+    std::unordered_set<std::string> excluded_providers;
 #if defined(OPENVINO_CONFIG_GPU_FP16) || defined(OPENVINO_CONFIG_GPU_FP32)
-    test.ConfigExcludeEps({kOpenVINOExecutionProvider});  // OpenVINO: Temporarily disabled due to accuracy issues
+    excluded_providers.insert(kOpenVINOExecutionProvider);  // OpenVINO: Temporarily disabled due to accuracy issues
 #endif
-    test.Config(run_with_tunable_op)
+#if defined(USE_QNN)
+    if (b_is_initializer && !c_is_initializer) {
+      // Accuracy issues on QNN's CPU backend with QNN SDK version 2.17
+      excluded_providers.insert(kQnnExecutionProvider);
```
Comment (adrianlizarraga) [inline onnxruntime/test/providers/cpu/math/gemm_test.cc:368]: Note: added the test `QnnCPUBackendTests.DISABLED_Gemm_Broadcast_Bias_DynamicA_StaticB_DynamicC` with a note to enable here when fixed.

File: onnxruntime/test/providers/cpu/math/matmul_test.cc [hunk 5]
```diff
@@ -173,6 +173,12 @@ void RunMatMulTest(int32_t opset_version, bool is_a_constant, bool is_b_constant
       // QNN can't handle 0 shap
       excluded_providers.insert(kQnnExecutionProvider);
     }
+#if defined(__linux__)
+    if (t.name == "test padding and broadcast B > A") {
+      // Accuracy error with QNN SDK 2.17.0 on CPU backend.
+      excluded_providers.insert(kQnnExecutionProvider);
```
Comment (adrianlizarraga) [inline onnxruntime/test/providers/cpu/math/matmul_test.cc:179]: Note: added the test `QnnCPUBackendTests.DISABLED_MatMulOp_PaddingAndBroadcast_BLargerThanA` with a note to re-enable this when fixed.

File: onnxruntime/test/providers/qnn/qnn_test_utils.h [hunk 6]
```diff
@@ -239,6 +249,19 @@ void InferenceModel(const std::string& model_data, const char* log_id,
  */
 void TryEnableQNNSaver(ProviderOptions& qnn_options);
 
+struct QDQTolerance {
+  // When comparing output activations between QNN EP and CPU EP (both running the QDQ model),
+  // this value defines the maximum tolerable difference as a percentage of the output range.
+  // Ex: (qdq@QNN_EP - qdq@CPU_EP) / (rmax_output - rmin_output) <= DEFAULT_QDQ_TOLERANCE.
+  static constexpr float DEFAULT_QDQ_TOLERANCE = 0.004f;  // 0.4% is equivalent to 1 int8 quantization unit
+                                                          // or 262 int16 quantization units.
```
Comment (adrianlizarraga) [inline onnxruntime/test/providers/qnn/qnn_test_utils.h:257]: Note: previous used 1 quantization unit + epsilon as the allowed tolerance between qdq model results. Now, reframing this tolerance as a percentage of the output range. 1 quantization unit is 1/255 = 0.4% for 8bit.

**Review by jywu-msft — APPROVED** [review]

**Review by HectorSVC — APPROVED** [review]

**Pull request merged**